File: src/main/java/io/github/justfoxx/teacup/v1/registry/MapRegistry.java
package io.github.justfoxx.teacup.v1.registry;

import com.google.common.collect.ImmutableSet;
import oshi.util.tuples.Pair;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * A registry that uses a {@link ConcurrentHashMap}
 * @param <K> The key type
 * @param <V> The value type
 */
public class MapRegistry<K,V> {
    protected final Map<K,V> registry = new ConcurrentHashMap<>();

    /**
     * Adds a key and value to the registry
     * @param key The key to add
     * @param value The value to add
     * @return A pair of the key and value
     */
    public Pair<K, V> add(K key, V value) {
        Objects.requireNonNull(value);
        Objects.requireNonNull(key);
        registry.putIfAbsent(key, value);
        return new Pair<>(key, value);
    }

    /**
     * Use {@link MapRegistry#getOptional(K)} instead
     * @param key The key to get
     * @return The value associated with the key
     */
    public V get(K key) {
        return registry.get(key);
    }

    /**
     * Gets a value from the registry
     * @param key The key to get
     * @return An optional of the value associated with the key
     */
    public Optional<V> getOptional(K key) {
        Objects.requireNonNull(key);
        return Optional.ofNullable(get(key));
    }

    /**
     * Gets all the entries in the registry
     * @return All the entries in the registry
     */
    public ImmutableSet<Map.Entry<K,V>> getAll() {
        return ImmutableSet.copyOf(registry.entrySet());
    }
}

File: src/main/kotlin/io/github/justfoxx/teacup/v1/registry/MapRegistry.kt
package io.github.justfoxx.teacup.v1.registry

import oshi.util.tuples.Pair
import java.util.concurrent.ConcurrentHashMap

/**
 * A thread-safe registry that stores key-value pairs.
 */
class MapRegistry<K, V> {

    // Use ConcurrentHashMap to ensure thread-safety
    private val registry: MutableMap<K, V> = ConcurrentHashMap()

    /**
     * Adds a key-value pair to the registry.
     *
     * @param key The key to add.
     * @param value The value to add.
     * @return A Pair of the added key and value.
     */
    fun add(key: K, value: V): Pair<K, V> {
        registry.putIfAbsent(key, value)
        return Pair(key, value)
    }

    /**
     * Gets the value associated with the given key from the registry.
     *
     * @param key The key to get the value for.
     * @return The value associated with the key, or null if the key is not found in the registry.
     */
    operator fun get(key: K): V? {
        return registry[key]
    }

    /**
     * Gets an immutable set of all key-value pairs in the registry.
     *
     * @return An immutable set of all key-value pairs in the registry.
     */
    val all: Set<Map.Entry<K, V>>
        get() = registry.entries.toSet()
}